File: src/agentswarm/adapters/placeholders.py
"""Placeholder expansion for adapter command templates.

An adapter command is a list of argument templates such as
``["codex", "exec", "--cd", "{workspace}", "-"]``. Before a command runs, every
``{name}`` token is replaced with a value drawn from the call context. Keeping
this as a small pure function (no I/O, no globals) makes the substitution rules
trivial to read and to unit test in isolation.
"""

from __future__ import annotations

import re

# The placeholders an adapter command (or stdin template) may reference.
PLACEHOLDERS = ("prompt", "prompt_file", "workspace", "source", "adapter", "role")

_TOKEN = re.compile(r"\{(" + "|".join(PLACEHOLDERS) + r")\}")


def expand(template: str, context: dict[str, str]) -> str:
    """Replace every known ``{placeholder}`` in *template* using *context*.

    Unknown ``{...}`` tokens are left untouched so that literal braces in a
    command (for example a shell brace expansion) survive. A known placeholder
    that is missing from *context* expands to an empty string.
    """
    return _TOKEN.sub(lambda m: context.get(m.group(1), ""), template)


def expand_all(templates: list[str], context: dict[str, str]) -> list[str]:
    """Expand a list of argument templates."""
    return [expand(t, context) for t in templates]

File: src/adapters/placeholders.ts
/**
 * Placeholder expansion for adapter command templates.
 *
 * An adapter command is a list of argument templates such as
 * `["codex", "exec", "--cd", "{workspace}", "-"]`. Before a command runs, every
 * known `{name}` token is replaced with a value from the call context. Unknown
 * `{...}` tokens are left untouched so literal braces survive; a known token
 * missing from the context expands to an empty string.
 */

export const PLACEHOLDERS = [
  "prompt",
  "prompt_file",
  "workspace",
  "source",
  "adapter",
  "role",
] as const;

export type PlaceholderName = (typeof PLACEHOLDERS)[number];
export type PlaceholderContext = Partial<Record<PlaceholderName, string>>;

const TOKEN = new RegExp(`\\{(${PLACEHOLDERS.join("|")})\\}`, "g");

/** Replace every known `{placeholder}` in `template` using `context`. */
export function expand(template: string, context: PlaceholderContext): string {
  return template.replace(TOKEN, (_match, name: PlaceholderName) => context[name] ?? "");
}

/** Expand a list of argument templates. */
export function expandAll(templates: string[], context: PlaceholderContext): string[] {
  return templates.map((t) => expand(t, context));
}
